### ISSUE TITLE
[ENH] design question: what should clusterers with `capability:out_of_sample = False` do when `predict` sees new data?

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#7594]
Issue for design discussion: what should clusterers with `capability:out_of_sample = False` do when `predict` sees new data?

The general issue is, clusterers have a taggable property called `capability:out_of_sample` - if this is `False`, the `X` in `predict` must be the same as in `fit`. This property does not occur for classifiers or regressors. Both the `True` and the `False` case are common.

The specific question is what to do if the `X` in `predict` is different from the `X` seen in fit. Some options:

* refit a clone on the new data, return labels
* in `TimeSeriesDBScan` I have made a different choice, refit on pooled data, i.e., include the data seen in `fit` as well.
* warn and do nothing else
* raise and do nothing else

A secondary problem is the necessary memory and compute footprint to determine in `predict` whether the data has already been seen. If we check identity every time, this requires memorization of the data in `fit` (or hashing), and a check in `predict`, which may be subtantial overhead. In https://github.com/sktime/sktime/pull/7593, I suggest to hash a printout but not more, to avoid the memory issues we have on forecasting (see https://github.com/sktime/sktime/issues/7416)

**Comment by yarnabrina** [comment]
I'll request to consider a new option, similar to how non-time-series clustering models work.

1. Take X in predict.
2. For series present in that X, find the distance from clusters based on trained data.
3. Return a series of labels.

So in this case, no need to memorise X. If one does, that can add the benefit of not calculating the distances again.

If I understand you correctly, it does not fall under none of the 4 aforementioned options.

These are my main reasons (especially point 3) for not liking above 4:

1. Even if a strict subset of X is passed, it will be considered not predictable as is.
2. Even if the full X is passed but by reordering the hierarchies (permutation of different series in X, keeping sequence for each series intact), it will still not work.
3. In my opinion, predict should not modify the state of the instance or any properties. If we have a tag specifically suggesting that, I'll consider it to be a very different from usual sklearn archtecture and possibly misleading.

As said in #7592, I do not use time series clustering myself, so mentioning that as a disclaimer to avoid affecting clustering standards too much.

**Comment by fkiraly** [comment]
@yarnabrina, I do agree with your feeling that the solutions are unsatisfactory, but I cannot come up with a better one that is easy to implement (and does not introduce memory usage or similar).

> If I understand you correctly, it does not fall under none of the 4 aforementioned options.

Yes - this is, in my opinion, not a general option, because the "distance from the cluster" is ill-defined. Even if there is a cluster center (which not all algorithms produce!), this implies a choice of distance, and not all clutsering algorithms are distance based.

> These are my main reasons (especially point 3) for not liking above 4:
> 1. Even if a strict subset of X is passed, it will be considered not predictable as is.
> 2. Even if the full X is passed but by reordering the hierarchies (permutation of different series in X, keeping sequence for each series intact), it will still not work.

Agreed, I do not like this either.

>3. In my opinion, predict should not modify the state of the instance or any properties.

It does not, in either of the options - it fits a clone which is then discarded.

> If we have a tag specifically suggesting that, I'll consider it to be a very different from usual sklearn archtecture and possibly misleading.

The option that sklearn takes is simply not having a `predict`, and only `fit_predict` - this is perhaps another option, although it would necessitate a `fit_predict_proba` because our clusterers all have `predict_proba`.

Example:
https://scikit-learn.org/1.6/modules/generated/sklearn.cluster.DBSCAN.html

> but the check must be more thorough allowing points 1/2 (subset/reorder in my reasons above) since those are not out-of-sample.

My main concern is, are we inviting a compute or memory leak? If the inner algorithm does `fit_predict` only, we have to run a generic algorithm to find the permutation or subset map that samples in your scenarios 1/2 correspond to.

Effective algorithms for "check whether set B is contained in set A" involves sorting or hashing, which are unpleasant operations to implement for rows of `pandas` `DataFrames` or similar - and note that we support many input types, which makes this check even more unpleasant if we want it on boilerplate level.
